### ISSUE TITLE
It is not possible to have a text node (span) with the same format as its grandparent node

### ISSUE THREAD

**Issue by dgmlr** [tinymce/tinymce#5826]
**What is the current behavior? Describe the bug** 

It is not possible to have a text node (span) with the same format as its grandparent node.

**Please provide the steps to reproduce and if possible a minimal demo of the problem via [fiddle.tinymce.com](http://fiddle.tinymce.com/) or similar.**

Consider this example configuration:

```
tinymce.init({
    selector: "textarea",
    formats: {
        x: {inline: 'span', classes: 'x', exact: true },
        y: {inline: 'span', classes: 'y', exact: true }},
    style_formats: [
        {title: 'x', format: 'x'},
        {title: 'y', format: 'y'}],
    content_style: 'span.x {color: blue} span.y {color: red}',
```
Then it is not possible to create a structure such as this

```
<span class="x"> 
one 
    <span class="y">
    two 
        <span class="x">
        three
        </span>
    four
    </span>
five
</span>
```

Whichever order the formats are applied in, the innermost span is automatically removed, leaving this structure:

```
<span class="x"> 
one 
    <span class="y">
    two three four
    </span>
five
</span>
```
There appears to be an assumption that the innermost span is redundant because it has the same format as the outermost span. But this assumption is incorrect, as can been seen from the above example; the word "three" will be rendered in the wrong color.

**What is the expected behavior?**

Spans should not be removed based on the format of the grandparent node (or ancestor of the grandparent). 

**Which versions of TinyMCE, and which browser / OS are affected by this issue?**

I've tested this with TinyMCE 5.3.2 and Chrome 80.0.3987.132

**Comment by lostkeys** [comment]
You are correct. TinyMCE does not nest styles. Instead it breaks up the spans. Using your example I end up with the following code which is expected of TinyMCE. [See an example here](http://fiddle.tinymce.com/Pmhaab/2)
```
<p>
    <span class="x">
        one 
        <span class="y">
            two 
        </span>                
        three
        <span class="y">
            four
        </span>
        five
    </span>
</p>
```
Which gives you
<img width="213" alt="image" src="https://user-images.githubusercontent.com/3122929/85998336-f15c0380-ba0a-11ea-93a5-1630d22b891d.png">

I can't replicate your issue where "three" is red.

I'm closing this issue as TinyMCE works as expected. Please feel free to reopen the issue if you have additional details

**Comment by dgmlr** [comment]
Yes, you can get the same visual result with a different HTML structure, but that is not related to the bug I'm reporting.
Here is a more specific set of steps to reproduce the problem:

1. Select the text "three" and apply format "x".
2. Select the text "two three four" and apply format "y".
3. Select the text "one two three four five" and apply format "x". Note that format "x" applied at step 1 is now gone (and hence "three" is now displaying the wrong color). Information has been lost, hence why this is a bug.

Note that in this scenario TinyMCE is _not_ splitting any span as you suggest, it is just deleting one of them. But nor would I agree that potentially splitting spans to maintain a visual result is a good solution anyway, as this is still deleting information by altering the semantic structure of the nested spans.  TinyMCE should just leave alone the nesting structure intended by the user and leave the visual result to CSS.

**Comment by lostkeys** [comment]
I agree, it should retain the inner span. In what use case did you discover this issue?

**Comment by lostkeys** [comment]
[Updated fiddle](http://fiddle.tinymce.com/Pmhaab/3). Select all text and apply the blue x format. The expected result is that the existing x format should remain.

**Comment by dgmlr** [comment]
Thanks, Fredrik.

My use case? Yes, that is not obvious from my highly simplified example.

What I've been working on is a notation for explaining the solution to cryptic crossword clues. This can involve complex transformations of components of the clue into components of the answer, including synonyms, abbreviations, reversals, anagrams, insertions, and translations.  My notation utilises multiple colors, underlines, overlines, bold, and italics to show these different components based on nested semantic span tags.  I've set up TinyMCE to create/edit these spans (because editing the HTML directly is ... painful).  Mostly this has been working really well.

Here is the example where I ran into the bug. I had previously created this by hand, but then found I couldn't edit it in TinyMCE without losing part of it.

```
<span class="combination">
    <span class="extraction">
        <span class="source">
            <span class="letters">
            C
            </span>
            <span class="combination">
                 <span class="sourcegroup">
                     <span class="letters">
                     A S S
                     </span>
                     <span class="letters">
                     A N D
                     </span>
                </span>
            </span> 
            <span class="letters">
            R
            </span>
        </span>
    </span> 
    <span class="letters">
    A
    </span>
</span>
```

**Comment by lostkeys** [comment]
Very cool use case indeed, thanks for sharing! We're investigating if this is truly is a bug or if it was intentionally written like this to avoid some other issues. We will get back to you.

Fyi, I tried another solution using custom elements which works the way you described. http://fiddle.tinymce.com/Pmhaab/4.

**Comment by dgmlr** [comment]
Thanks Frederik. To give your more context about the use case, here are some examples of how the nested spans are rendered (with my current work-in-progress CSS):
![nested spans use case examples](https://user-images.githubusercontent.com/67579927/86091213-d6fb5600-baee-11ea-921a-3f22cc30aca9.PNG)
